File: Data Structures/Trees/Depth-First Search/94/94.go
package main

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// inorderTraversal performs an in-order traversal of a binary tree and returns the values of the nodes in the tree in an
func inorderTraversal(root *TreeNode) []int {
	// If the root is nil, return an empty slice
	if root == nil {
		return []int{}
	}
	// Create an empty slice to store the inorder traversal
	inorder := []int{}
	// Perform an in-order traversal and append the values to the slice
	inorder = append(inorder, inorderTraversal(root.Left)...)
	inorder = append(inorder, root.Val)
	inorder = append(inorder, inorderTraversal(root.Right)...)
	// Return the inorder traversal
	return inorder
}

// time complexity: O(n)
// space complexity: O(n)

File: Data Structures/Trees/Depth-First Search/94/94.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        """
        Perform an inorder traversal of a binary tree starting from the given root node.

        Parameters:
            root (Optional[TreeNode]): The root node of the binary tree.

        Returns:
            List[int]: A list containing the values of the nodes in inorder traversal.
        """
        result = []

        def helper(root):
            if not root:
                return
            helper(root.left)
            result.append(root.val)
            helper(root.right)
        
        helper(root)
        return result

# Time Complexity: O(N)
# Space Complexity: O(N)